let show_plot_2=true;let plot_2={"$schema":"https://vega.github.io/schema/vega-lite/v4.json","width":"container","layer":[{"data":{"values":[{"key":"Characterizing the molecular regulation of inhibitory immune checkpoints with multimodal single-cell screens","value":4},{"key":"High-throughput single-cell chromatin accessibility CRISPR screens enable unbiased identification of regulatory networks in cancer\r","value":3},{"key":"Mapping information-rich genotype-phenotype landscapes with genome-scale Perturb-seq","value":3},{"key":"A multiplexed single-cell CRISPR screening platform enables systematic dissection of the unfolded protein response","value":3},{"key":"Massively multiplex chemical transcriptomics at single-cell resolution","value":3},{"key":"A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens","value":3},{"key":"Genome-wide CRISPRi/a screens in human neurons link lysosomal failure to ferroptosis","value":2},{"key":"Profiling the genetic determinants of chromatin accessibility with scalable single-cell CRISPR screens\n ","value":2},{"key":"Multimodal pooled Perturb-CITE-seq screens in patient models define mechanisms of cancer immune evasion","value":2},{"key":"Optimal-Transport Analysis of Single-Cell Gene Expression Identifies Developmental Trajectories in Reprogramming","value":2}]},"mark":"bar","encoding":{"x":{"field":"key","sort":{"field":"value","order":"descending"},"title":"Title"},"y":{"field":"value","type":"quantitative","title":null}}}]}